### ISSUE TITLE
Deserialize HAL responses

### ISSUE THREAD

**Issue by viesti** [cognitect-labs/aws-api#30]
Seems that API Gateway responses are not deserialized currently. I don't know the proper place/way to do deserialization, but here might be a pointer do get the embedded data of the HAL response: https://github.com/portkey-cloud/aws-clj-sdk/blob/master/src/portkey/aws.clj#L182-L197.

**Comment by dchelimsky** [comment]
The whole response is not deserialized, or do you mean the body? Can you post what you're getting back and what you're expecting to get back?

**Comment by viesti** [comment]
Ah sorry, forgot to post a sample :(

```clojure
user=> (aws/invoke apigw-client {:op :GetRestApis})
{}
user=> (-> (aws/invoke apigw-client {:op :GetRestApis}) meta :http-response :body slurp json/read-json keys)
(:_links :_embedded)
```

The HTTP response body in the metadata is a JSON string in HAL form. I guess it could be deserialized as the result of `aws/invoke`.

**Comment by dchelimsky** [comment]
Thanks for that additional detail. I expect the same thing you do!

FYI, you can also see the response structure like this:

``` clojure
 user=> (-> apigw-client aws/ops :GetRestApis :response)
 {:position string,
  :items
  [:seq-of
   {:endpointConfiguration  {:types [:seq-of string]},
    :description            string,
    :name                   string,
    :warnings               [:seq-of string],
    :apiKeySource           string,
    :policy                 string,
    :binaryMediaTypes       [:seq-of string],
    :id                     string,
    :createdDate            timestamp,
    :version                string,
    :minimumCompressionSize integer}]}
```

**Comment by viesti** [comment]
Had a quick look and found `client/parse-http-response` multimethod, which dispatches on the protocol name, which is `rest-json` for API Gateway. The content-type http header could further used for selecting how to parse the response:

```clojure
user> (-> apigw (aws/invoke {:op :GetRestApis}) meta (get-in [:http-response :headers "content-type"]))
"application/hal+json"
```

**Comment by dchelimsky** [comment]
Thanks @viesti - that's the same path I'm on. Just haven't had time to focus on this issue.

**Comment by viesti** [comment]
No problem, just had an inspiration to look into this. Focus&time are Muses that I find hard to catch :)

**Comment by dchelimsky** [comment]
Happy muse-hunting! And thanks for checking in. This just might get some attention now ;)

**Comment by dchelimsky** [comment]
So I started to look into this and discovered (and fixed) an edge-case bug parsing json structures. This is fixed on master, and was necessary for me to move forward w/ the HAL issue. Will keep you posted.

**Comment by dchelimsky** [comment]
I've got a solution for this on the `issue-30-hal-responses`branch, but I have a few things I want to work through in the code before I merge it.

**Comment by viesti** [comment]
Nice! Good thing if this work generally improves the library :)

**Comment by viesti** [comment]
True, I found that out myself only recently.

**Comment by dchelimsky** [comment]
Thanks for the feedback @maxthoursie. I'm aiming to get a release out today.

**Comment by dchelimsky** [comment]
Fix merged to master. Release later today.

**Comment by dchelimsky** [comment]
Released in 0.8.198.
